src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (let i = 0; i < actions.length; i++) {
+    const action = actions[i];
+    let nextState = { ...currentState };
+
+    switch (action.type) {
+      case 'clear':
+        nextState = {};
+        break;
+      case 'addProperties':
+        Object.assign(nextState, action.extraData);
+        break;
+      case 'removeProperties':
+        action.keysToRemove.forEach((key) => delete nextState[key]);
+        break;
+      default:
+        break;
+    }
+
+    stateHistory.push(nextState);
+    currentState = nextState;
```

Comment:
You should avoid reassigning `currentState` to `nextState` at the end of each loop iteration. Instead, make a copy of the state at the beginning of the next iteration. This will prevent potential issues with object references and ensure that `currentState` always refers to the state before the current action.